### PR TITLE
Expose internals to support randomization (simplified)

### PR DESCRIPTION
**Pull request by conradoplg**
Part of https://github.com/ZcashFoundation/frost/issues/121 along with https://github.com/ZcashFoundation/reddsa/pull/33

This is alternative to https://github.com/ZcashFoundation/frost/pull/83, which requires less change but is conceptually less "pure". The main thing that is required for randomized FROST is having the randomized group public key hashed when computing the binding factors. This introduces an internal function with a "additional_prefix" parameter that is empty in regular FROST and can be used to randomized FROST.

This allows less duplicated code in randomized FROST (basically just the high-level `sign` and `aggregate` functions).

While targeting `main`, this builds on top of https://github.com/ZcashFoundation/frost/pull/83, so you can see the last commit to check the differences between those.

### CODE REVIEW

File: frost-core/src/frost.rs [hunk 1]
```diff
@@ -255,43 +279,41 @@ pub struct GroupCommitment<C: Ciphersuite>(pub(super) Element<C>);
...
 
-impl<C> TryFrom<&SigningPackage<C>> for GroupCommitment<C>
+/// Generates the group commitment which is published as part of the joint
+/// Schnorr signature.
+///
+/// Implements [`compute_group_commitment`] from the spec.
+///
+/// [`compute_group_commitment`]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-10.html#section-4.5
+#[cfg_attr(feature = "internals", visibility::make(pub))]
+fn compute_group_commitment<C>(
```
Comment (dconnolly) [inline frost-core/src/frost.rs:289]: Can we also call this function from the `TryFrom` impl for the default (`internals` feature not enabled) case?

File: frost-core/src/frost/round2.rs [hunk 2]
```diff
@@ -109,6 +109,25 @@ where
...
+    challenge: Challenge<C>,
+) -> SignatureShare<C> {
+    let z_share: <<C::Group as Group>::Field as Field>::Scalar = signer_nonces.hiding.0
+        + (signer_nonces.binding.0 * binding_factor.0)
+        + (lambda_i * key_package.secret_share.0 * challenge.0);
+    let signature_share = SignatureShare::<C> {
+        identifier: *key_package.identifier(),
+        signature: SignatureResponse::<C> { z_share },
+    };
+    signature_share
```
Comment (dconnolly) [inline frost-core/src/frost/round2.rs:128]:
```suggestion
    let z_share: <<C::Group as Group>::Field as Field>::Scalar = signer_nonces.hiding.0
        + (signer_nonces.binding.0 * binding_factor.0)
        + (lambda_i * key_package.secret_share.0 * challenge.0);
        
    return SignatureShare::<C> {
        identifier: *key_package.identifier(),
        signature: SignatureResponse::<C> { z_share },
    }
```

File: frost-core/src/signature.rs [hunk 3]
```diff
@@ -20,6 +20,15 @@ where
...
     <C::Group as Group>::Field: Field,
 {
+    /// Create a new Signature.
+    #[cfg(feature = "internals")]
+    pub fn new(
+        R: <C::Group as Group>::Element,
+        z: <<C::Group as Group>::Field as Field>::Scalar,
+    ) -> Self {
+        Self { R, z }
+    }
```
Comment (dconnolly) [inline frost-core/src/signature.rs:30]: 👍

File: frost-core/src/verifying_key.rs [hunk 4]
```diff
@@ -23,6 +23,18 @@ where
...
+    #[cfg(feature = "internals")]
+    pub fn new(element: <C::Group as Group>::Element) -> Self {
+        Self { element }
+    }
+
+    /// Return the underlying element.
+    #[cfg(feature = "internals")]
+    pub fn to_element(self) -> <C::Group as Group>::Element {
+        self.element
+    }
```
Comment (dconnolly) [inline frost-core/src/verifying_key.rs:36]: 👍

**Review by dconnolly — DISMISSED** [review]
looking good, some comments

File: frost-core/src/frost.rs (hunk 1 above)
Comment (conradoplg) [reply frost-core/src/frost.rs:289]: We can, but it wouldn't be used. `compute_group_commitment` receives a previously-computed `binding_factor_list` (which _is_ previously computed everywhere it's called) so there is no motivation to actually call the TryFrom (which would recompute `binding_factor_list`). We could make a `TryFrom` from a tuple, but I think that's usually non-intuitive...

File: frost-core/src/frost/round2.rs (hunk 2 above)
Comment (conradoplg) [reply frost-core/src/frost/round2.rs]: Done in 1b9efcf without the `return`

File: frost-rerandomized/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+/// Implements [`sign`] from the spec.
+///
+/// Receives the message to be signed and a set of signing commitments and a set
+/// of randomizing commitments to be used in that signing operation, including
+/// that for this participant.
+///
+/// Assumes the participant has already determined which nonce corresponds with
+/// the commitment that was assigned by the coordinator in the SigningPackage.
+///
+/// [`sign`]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-10.html#name-round-two-signature-share-g
```
Comment (dconnolly) [inline frost-rerandomized/src/lib.rs:28]:
```suggestion
/// [`sign`]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-11.html#name-round-two-signature-share-g
```

File: frost-core/src/frost.rs [hunk 6]
```diff
@@ -223,13 +232,18 @@ where
 
     /// Compute the preimages to H3 to compute the per-signer binding factors
```
Comment (dconnolly) [inline frost-core/src/frost.rs:233]: s/H3/H1/ ?

https://github.com/ZcashFoundation/frost/blob/a02649d75685c20d972d15ff47ff04084d03fd8a/frost-core/src/frost.rs#L123

```suggestion
    /// Compute the preimages to H1 to compute the per-signer binding factors
```

File: frost-rerandomized/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        &self.randomizer_point
+    }
+
+    /// Return the randomized group public key.
+    ///
+    /// It can be used to verify the final signature.
+    pub fn randomized_group_public_key(&self) -> &frost_core::VerifyingKey<C> {
+        &self.randomized_group_public_key
+    }
+}
```
Comment (dconnolly) [inline frost-rerandomized/src/lib.rs:209]: 👍

**Review by dconnolly — DISMISSED** [review]
lgtm! very nice

**Review by dconnolly — APPROVED** [review]
lgtm! very nice

**Pull request merged**